In [2]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler, StandardScaler

Reading Movie meta data

In [3]:
movie_data=pd.read_csv('../../Data_Files/movie_metadata.csv')
movie_data.shape

(5043, 28)

### How to tackle Categorical Data
So their are 2 ways we can tackle categorical data
1. One-Hot Incoding
2. Label Incoding

1. One-Hot Incoding :  here all unique data indexs become a columns with binary data representation like (0/1), where it create a maxtrix with 0/1 where 1 represent the data while 0 represent not entry

2. Label Incoding : here all inputs are get assign a uniqe index and arrage in alfabetical order in single column


### How to convert Categorical data into Numerical Data for better comparision

In [4]:
titanic_data=pd.read_csv('../../Data_Files/Titanic.csv')
titanic_data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [5]:
titanic_data.shape

(891, 12)

In [6]:
titanic_data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [7]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Types of categorical Variable : 
1. Nominal - no numerical data and no order (No heirachy)
2. Ordinal - no numrical data, but order present (heirarchy present)
3. Binary - having only 2 values (Male/Female)

### Que1. How many male and female passenger are in ship ?

In [11]:
titanic_data['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

=> This shows thier are more male passenger as compare to female 

### Que2. How many people service and not servive ?

In [12]:
titanic_data['Survived'].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

0 - not servived (549)
1 - Servived (342)

### Que3 . How many male and how many female pasenger servived and not servived ?

In [18]:
titanic_data.groupby('Survived').agg({'Sex' : 'value_counts'})

Sex
Survived Sex        
0        male    468
         female   81
1        female  233
         male    109

So we can see out of female - females are survived more than male passengers

### One-Hot Incoding for Normalisation and Standardisation - to convert categorical data into Numerical data

1. Pandas df provide one function - pd.get_dummies(data, column)
2. Here in One-Hot incoding - every unique data get one idividual column so n unique values in categorical column we get n unique coumns with their name
3. For Algorithm, feading numerical values is much more easier and simpler as compare to feading categorical data so for this purpose, we convert categorical data into numerical data
4. True -mean 1 and False means 0

In [23]:
new_df = pd.get_dummies(titanic_data, columns=['Sex','Embarked']) # Sex_female & Sex_male as true/false representation while your original data is secure

In [26]:
new_df # Embarked_C/Q/S Sex_Female/Male

,PassengerId,Survived,Pclass,Name,Age,SibSp,Parch,Ticket,Fare,Cabin,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S
0,1,0,3,"Braund, Mr. Owen Harris",22.0,1,0,A/5 21171,7.2500,NaN,False,True,False,False,True
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1,0,PC 17599,71.2833,C85,True,False,True,False,False
2,3,1,3,"Heikkinen, Miss. Laina",26.0,0,0,STON/O2. 3101282,7.9250,NaN,True,False,False,False,True
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,0,113803,53.1000,C123,True,False,False,False,True
4,5,0,3,"Allen, Mr. William Henry",35.0,0,0,373450,8.0500,NaN,False,True,False,False,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",27.0,0,0,211536,13.0000,NaN,False,True,False,False,True
887,888,1,1,"Graham, Miss. Margaret Edith",19.0,0,0,112053,30.0000,B42,True,False,False,False,True
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",NaN,1,2,W./C. 6607,23.4500,NaN,True,False,False,False,True
889,890,1,1,"Behr, Mr. Karl Howell",26.0,0,0,111369,30.0000,C148,False,True,True,False,False


### Import LabelIncoder package from sklearn.preprocessing
1. Initialise the variable with LabelIncoder() functions
2. Then apply label icoded technique to create new columns in dataet using that variable
3. New column get create with numerical representation - with alfabetical order

In [28]:
from sklearn.preprocessing import LabelEncoder

In [29]:
label_incoder=LabelEncoder()

In [30]:
titanic_data['Incoded_Gender']= label_incoder.fit_transform(titanic_data['Sex'])

In [31]:
titanic_data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Incoded_Gender
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,1
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S,0
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S,1


### What if dataset have mix of both numerical and Alfabetical data
Then you can use both as it will treated as string accordingly it will treat

In [33]:
titanic_data['Incoded_Embarked']=label_incoder.fit_transform(titanic_data['Embarked'])

In [34]:
titanic_data

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Incoded_Gender,Incoded_Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,1,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,2
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S,0,2
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S,1,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S,1,2
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S,0,2
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S,0,2
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C,1,0


In [37]:
titanic_data['Incoded_Embarked'].value_counts()

Incoded_Embarked
2    644
0    168
1     77
3      2
Name: count, dtype: int64

In [36]:
titanic_data['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

1. Here Embarked col have 2 nan values so first fill or remove those values and then convert that col into numrical data 
2. In real life - we need to use our intutive to take dision whether which data is important to use for analysis 